File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/ResponseHeadersBuilder.kt
package io.envoyproxy.envoymobile

/*
 * Builder used for constructing instances of `ResponseHeaders`.
 */
class ResponseHeadersBuilder : HeadersBuilder {

  /**
   * Initialize a new instance of the builder.
   */
  constructor() : super(mutableMapOf())

  /**
   * Instantiate a new builder. Used only by ResponseHeaders to convert back to
   * ResponseHeadersBuilder.
   *
   * @param headers: The headers to start with.
   */
  internal constructor(headers: MutableMap<String, MutableList<String>>) : super(headers)

  override fun add(name: String, value: String): ResponseHeadersBuilder {
    super.add(name, value)
    return this
  }

  override fun set(name: String, value: MutableList<String>): ResponseHeadersBuilder {
    super.set(name, value)
    return this
  }

  override fun remove(name: String): ResponseHeadersBuilder {
    super.remove(name)
    return this
  }

  override fun internalSet(name: String, value: MutableList<String>): ResponseHeadersBuilder {
    super.internalSet(name, value)
    return this
  }

  /**
   * Add an HTTP status to the response headers.
   *
   * @param status: The HTTP status to add.
   *
   * @return ResponseHeadersBuilder, This builder.
   */
  fun addHttpStatus(status: Int): ResponseHeadersBuilder {
    internalSet(":status", mutableListOf("$status"))
    return this
  }

  /**
   * Build the response headers using the current builder.
   *
   * @return ResponseHeaders, New instance of response headers.
   */
  fun build(): ResponseHeaders {
    return ResponseHeaders(headers)
  }
}

File: mobile/library/swift/src/ResponseHeadersBuilder.swift
import Foundation

/// Builder used for constructing instances of `ResponseHeaders`.
@objcMembers
public final class ResponseHeadersBuilder: HeadersBuilder {
  /// Initialize a new instance of the builder.
  public override convenience init() {
    self.init(headers: [:])
  }

  /// Add an HTTP status to the response headers.
  ///
  /// - parameter status: The HTTP status to add.
  ///
  /// - returns: This builder.
  public func addHttpStatus(_ status: Int) -> ResponseHeadersBuilder {
    self.internalSet(name: ":status", value: ["\(status)"])
    return self
  }

  /// Build the response headers using the current builder.
  ///
  /// - returns: New instance of response headers.
  public func build() -> ResponseHeaders {
    return ResponseHeaders(headers: self.headers)
  }
}
